`paramListToForm` fails to build with `http-api-data` >= 0.7

## Description

  Scotty 0.30 fails to build when the solver picks `http-api-data >= 0.7`, which changed the `Form` representation from `HashMap Text [Text]` to `Map Text [Text]` (for HashDoS mitigation).

  ## Error

  Web/Scotty/Action.hs:459:97: error:
      • Couldn't match expected type: Map.Map T.Text [T.Text]
                    with actual type: HM.HashMap k0 v0
      • In the second argument of 'foldl'', namely 'HM.empty'
        In the second argument of '(.)', namely
          'foldl' (\ f (k, v) -> HM.alter (prependValue v) k f) HM.empty'

  ## Cause

  `paramListToForm` uses `HM.empty` and `HM.alter` but the `Form` constructor now wraps `Map.Map
  T.Text [T.Text]`:

  ```haskell
  paramListToForm = Form . fmap reverse . foldl' (\f (k, v) -> HM.alter (prependValue v) k f) HM.empty
  ```

  ## Fix

  Replace the `HashMap` operations with `Map` equivalents:

  ```haskell
  paramListToForm = Form . fmap reverse . foldl' (\f (k, v) -> Map.alter (prependValue v) k f) Map.empty
  ```

  And update the imports accordingly (`Data.Map.Strict` instead of / in addition to `Data.HashMap.Strict`).

--- a/.github/workflows/haskell-ci.yml
+++ b/.github/workflows/haskell-ci.yml
@@ -24,6 +24,7 @@ jobs:
           - { cabal: "3.10", os: ubuntu-latest,  ghc: "9.8.4"  }
           - { cabal: "3.12", os: ubuntu-latest,  ghc: "9.10.3" }
           - { cabal: "3.14", os: ubuntu-latest,  ghc: "9.12.2" }
+          - { cabal: "3.16", os: ubuntu-latest,  ghc: "9.14.1" }
       fail-fast: false
 
     steps:

--- a/Web/Scotty/Action.hs
+++ b/Web/Scotty/Action.hs
@@ -452,7 +452,7 @@ formData = do
   where
     -- This rather contrived implementation uses cons and reverse to avoid
     -- quadratic complexity when constructing a Form from a list of Param.
-    -- It's equivalent to using HashMap.insertWith (++) which does have
+    -- It's equivalent to using Map.insertWith (++) which does have
     -- quadratic complexity due to appending at the end of list.
     paramListToForm :: [Param] -> Form
     paramListToForm = Form . fmap reverse . foldl' (\f (k, v) -> Map.alter (prependValue v) k f) Map.empty

--- a/Web/Scotty/Action.hs
+++ b/Web/Scotty/Action.hs
@@ -80,7 +80,6 @@ import qualified Data.ByteString.Lazy.Char8 as BL
 import qualified Data.CaseInsensitive       as CI
 import           Data.Traversable (for)
 import qualified Data.Map.Strict            as Map
-import qualified Data.HashMap.Strict        as HM
 import           Data.Int
 import           Data.List (foldl')
 import           Data.Maybe                 (maybeToList)
@@ -456,7 +455,7 @@ formData = do
     -- It's equivalent to using HashMap.insertWith (++) which does have
     -- quadratic complexity due to appending at the end of list.
     paramListToForm :: [Param] -> Form
-    paramListToForm = Form . fmap reverse . foldl' (\f (k, v) -> HM.alter (prependValue v) k f) HM.empty
+    paramListToForm = Form . fmap reverse . foldl' (\f (k, v) -> Map.alter (prependValue v) k f) Map.empty
 
     prependValue :: a -> Maybe [a] -> Maybe [a]
     prependValue v = Just . maybe [v] (v :)

--- a/changelog.md
+++ b/changelog.md
@@ -1,3 +1,8 @@
+## Next
+
+* Added support for GHC 9.14, containers-0.8 (a boot lib).
+* Raise lower bounds for a lot of dependencies
+
 ## 0.30 [2026.01.06]
 
 * Added sessions (#317).

--- a/changelog.md
+++ b/changelog.md
@@ -1,6 +1,8 @@
 ## Next
 
 * Added support for GHC 9.14, containers-0.8 (a boot lib).
+
+### Breaking changes
 * Raise lower bounds for a lot of dependencies
 
 ## 0.30 [2026.01.06]

--- a/scotty.cabal
+++ b/scotty.cabal
@@ -50,6 +50,7 @@ tested-with:         GHC == 8.10.7
                    , GHC == 9.8.4
                    , GHC == 9.10.3
                    , GHC == 9.12.2
+                   , GHC == 9.14.1
 Extra-source-files:
     README.md
     changelog.md
@@ -73,38 +74,32 @@ Library
                        Web.Scotty.Trans.Lazy
                        Web.Scotty.Util
   default-language:    Haskell2010
-  build-depends:       aeson                 >= 0.6.2.1  && < 2.3,
+  build-depends:       aeson                 >= 2.2.4.1  && < 2.3,
                        base                  >= 4.14     && < 5,
-                       blaze-builder         >= 0.3.3.0  && < 0.5,
-                       bytestring            >= 0.10.0.2 ,
-                       case-insensitive      >= 1.0.0.1  && < 1.3,
-                       containers            >= 0.5      && < 0.8,
-                       cookie                >= 0.4,
-                       exceptions            >= 0.7      && < 0.11,
-                       http-api-data         < 0.7,
-                       http-types            >= 0.9.1    && < 0.13,
-                       monad-control         >= 1.0.0.3  && < 1.1,
-                       mtl                   >= 2.1.2    && < 2.4,
-                       network               >= 2.6.0.2  && < 3.3,
+                       blaze-builder         >= 0.4.4.1  && < 0.5,
+                       bytestring            >= 0.10.12.0,
+                       case-insensitive      >= 1.2.1.0  && < 1.3,
+                       containers            >= 0.5      && < 0.9,
+                       cookie                >= 0.5.1,
+                       exceptions            >= 0.10.4   && < 0.11,
+                       http-api-data         >= 0.7,
+                       http-types            >= 0.12.4   && < 0.13,
+                       monad-control         >= 1.0.3.1  && < 1.1,
+                       mtl                   >= 2.2.2    && < 2.4,
+                       network               >= 3.2.8.0  && < 3.3,
                        regex-compat          >= 0.95.1   && < 0.96,
                        resourcet,
                        stm,
-                       text                  >= 0.11.3.1 ,
-                       time                  >= 1.8,
-                       transformers          >= 0.3.0.0  && < 0.7,
-                       transformers-base     >= 0.4.1    && < 0.5,
+                       text                  >= 1.2.4.1,
+                       time                  >= 1.9.3,
+                       transformers          >= 0.5.6.2  && < 0.7,
+                       transformers-base     >= 0.4.6.1  && < 0.5,
                        unliftio >= 0.2,
-                       unordered-containers  >= 0.2.10.0 && < 0.3,
-                       wai                   >= 3.0.0    && < 3.3,
-                       wai-extra             >= 3.1.14,
-                       warp                  >= 3.0.13,
-                       random                >= 1.0.0.0
-
-  if impl(ghc < 8.0)
-    build-depends:     fail
-
-  if impl(ghc < 7.10)
-    build-depends:     nats                  >= 0.1      && < 2
+                       unordered-containers  >= 0.2.21   && < 0.3,
+                       wai                   >= 3.2.4    && < 3.3,
+                       wai-extra             >= 3.1.18,
+                       warp                  >= 3.4.13,
+                       random                >= 1.3.1
 
   GHC-options: -Wall -fno-warn-orphans -Wno-unused-imports
 

